Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp

```diff
@@ -0,0 +1,280 @@
+                const PrefabDomValue* parentEntityDomInFocusedTemplate = entityPathInFocusedTemplate.Get(focusedTemplateDom);
+                AZ_Assert(parentEntityDomInFocusedTemplate,
+                    "Could not load parent entity's DOM from the focused template's DOM. "
+                    "Focused template id: '%llu'.", static_cast<AZ::u64>(m_templateId));
+
+                m_instanceToTemplateInterface->GeneratePatch(m_redoPatch,
+                    *parentEntityDomInFocusedTemplate, parentEntityDomAfterAddingEntity);
+
+                m_instanceToTemplateInterface->AppendEntityAliasPathToPatchPaths(m_redoPatch,
+                    parentEntityAliasPathForPatches);
+
+                if (updateUndoPatchNeeded)
+                {
+                    m_instanceToTemplateInterface->GeneratePatch(m_undoPatch,
+                        parentEntityDomAfterAddingEntity, *parentEntityDomInFocusedTemplate);
+
+                    m_instanceToTemplateInterface->AppendEntityAliasPathToPatchPaths(m_undoPatch,
+                        parentEntityAliasPathForPatches);
+                }
+            }
+        }
+
+        void PrefabUndoAddEntity::GeneratePatchesForAddNewEntity(
+            PrefabDom& newEntityDom,
+            const AZStd::string& newEntityAliasPathForPatches,
+            bool updateUndoPatchNeeded)
+        {
+            PrefabDomValue addNewEntityRedoPatch(rapidjson::kObjectType);
+            rapidjson::Value path = rapidjson::Value(newEntityAliasPathForPatches.data(),
+                aznumeric_caster(newEntityAliasPathForPatches.length()), m_redoPatch.GetAllocator());
+            rapidjson::Value patchValue;
+            patchValue.CopyFrom(newEntityDom, m_redoPatch.GetAllocator(), true);
+            addNewEntityRedoPatch.AddMember(rapidjson::StringRef("op"), rapidjson::StringRef("add"), m_redoPatch.GetAllocator())
+                .AddMember(rapidjson::StringRef("path"), AZStd::move(path), m_redoPatch.GetAllocator())
+                .AddMember(rapidjson::StringRef("value"), AZStd::move(patchValue), m_redoPatch.GetAllocator());
+            m_redoPatch.PushBack(addNewEntityRedoPatch.Move(), m_redoPatch.GetAllocator());
+
+            if (updateUndoPatchNeeded)
+            {
+                PrefabDomValue addNewEntityUndoPatch(rapidjson::kObjectType);
+                path = rapidjson::Value(newEntityAliasPathForPatches.data(),
+                    aznumeric_caster(newEntityAliasPathForPatches.length()), m_undoPatch.GetAllocator());
+                addNewEntityUndoPatch.AddMember(rapidjson::StringRef("op"), rapidjson::StringRef("remove"), m_undoPatch.GetAllocator())
+                    .AddMember(rapidjson::StringRef("path"), AZStd::move(path), m_undoPatch.GetAllocator());
+                m_undoPatch.PushBack(addNewEntityUndoPatch.Move(), m_undoPatch.GetAllocator());
+            }
+        }
+
+        void PrefabUndoAddEntity::GeneratePatchesForLinkUpdate()
+        {
```

Comment:
Can we reuse PrefabUndoLinkUpdate?
- If it is a source template manipulation, we do general undo/redo.
- If it is an override manipulation, we do link update.

I notice you already prepare the patches with override paths before the call. I think you can pass them to that undo node like this:

```cpp
PrefabUndoLinkUpdate* linkUpdate = aznew PrefabUndoLinkUpdate("Delete Override Link Update");
linkUpdate->SetParent(undoBatch.GetUndoBatch());
linkUpdate->Capture(overridePatches, instanceOverridePatchLinkId);
linkUpdate->Redo();
```